File: app/src/main/java/com/supercilex/robotscouter/util/ui/TeamMediaCreator.java
package com.supercilex.robotscouter.util.ui;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.content.FileProvider;
import android.widget.Toast;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.crash.FirebaseCrash;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.data.model.Team;
import com.supercilex.robotscouter.util.data.IoUtilsKt;

import java.io.File;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import pub.devrel.easypermissions.EasyPermissions;

import static com.supercilex.robotscouter.util.ConstantsKt.getProviderAuthority;
import static com.supercilex.robotscouter.util.data.IoUtilsKt.createFile;
import static com.supercilex.robotscouter.util.data.IoUtilsKt.getIO_PERMS;
import static com.supercilex.robotscouter.util.data.IoUtilsKt.getMediaFolder;

public final class TeamMediaCreator implements Parcelable, OnSuccessListener<Void>, ActivityCompat.OnRequestPermissionsResultCallback {
    public interface StartCaptureListener {
        void onStartCapture(boolean shouldUploadMediaToTba);
    }

    public static final Parcelable.Creator<TeamMediaCreator> CREATOR = new Parcelable.Creator<TeamMediaCreator>() {
        @Override
        public TeamMediaCreator createFromParcel(Parcel source) {
            return new TeamMediaCreator(source.readParcelable(Team.class.getClassLoader()),
                                        source.readString(),
                                        source.readInt() == 1);
        }

        @Override
        public TeamMediaCreator[] newArray(int size) {
            return new TeamMediaCreator[size];
        }
    };

    private static final int TAKE_PHOTO_RC = 334;
    private static final List<String> PERMS;
    private static final String MEDIA_CREATOR_KEY = "media_creator";

    static {
        List<String> perms = new ArrayList<>();
        perms.addAll(getIO_PERMS());
        perms.add(Manifest.permission.CAMERA);
        PERMS = Collections.unmodifiableList(perms);
    }

    private WeakReference<Fragment> mFragment;
    private Team mTeam;
    private PermissionRequestHandler mPermHandler;
    private WeakReference<OnSuccessListener<Team>> mListener;
    private String mPhotoPath;
    private boolean mShouldUploadMediaToTba;

    private TeamMediaCreator(Team team,
                             String photoPath,
                             boolean shouldUploadMediaToTba) {
        mTeam = team;
        mPhotoPath = photoPath;
        mShouldUploadMediaToTba = shouldUploadMediaToTba;
    }

    public static TeamMediaCreator newInstance(Fragment fragment,
                                               Team team,
                                               OnSuccessListener<Team> listener) {
        TeamMediaCreator mediaCreator = new TeamMediaCreator(team, null, false);
        init(mediaCreator, fragment, listener);
        return mediaCreator;
    }

    public static TeamMediaCreator get(Bundle bundle,
                                       Fragment fragment,
                                       OnSuccessListener<Team> listener) {
        TeamMediaCreator mediaCreator = bundle.getParcelable(MEDIA_CREATOR_KEY);
        init(mediaCreator, fragment, listener);
        return mediaCreator;
    }

    private static void init(TeamMediaCreator mediaCreator,
                             Fragment fragment,
                             OnSuccessListener<Team> listener) {
        mediaCreator.mFragment = new WeakReference<>(fragment);
        mediaCreator.mListener = new WeakReference<>(listener);
        mediaCreator.mPermHandler = new PermissionRequestHandler(PERMS, fragment, mediaCreator);
    }

    public Bundle toBundle() {
        Bundle args = new Bundle();
        args.putParcelable(MEDIA_CREATOR_KEY, this);
        return args;
    }

    public void setTeam(Team team) {
        mTeam = team;
    }

    @Override
    public void onSuccess(Void nothing) {
        startCapture(mShouldUploadMediaToTba);
    }

    public void startCapture(boolean shouldUploadMediaToTba) {
        mShouldUploadMediaToTba = shouldUploadMediaToTba;

        Context context = mFragment.get().getContext();
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if (takePictureIntent.resolveActivity(context.getPackageManager()) != null) {
            if (!EasyPermissions.hasPermissions(context, mPermHandler.getPermsArray())) {
                mPermHandler.requestPerms(R.string.media_write_storage_rationale);
                return;
            }

            File photoFile = null;
            try {
                //noinspection MissingPermission
                photoFile = createImageFile(getMediaFolder());
            } catch (Exception e) { // NOPMD
                FirebaseCrash.report(e);
                Toast.makeText(context, e.toString(), Toast.LENGTH_LONG).show();
            }

            if (photoFile != null) {
                mPhotoPath = photoFile.getAbsolutePath();

                Uri photoUri = FileProvider.getUriForFile(
                        context, getProviderAuthority(), photoFile);
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoUri);
                mFragment.get().startActivityForResult(takePictureIntent, TAKE_PHOTO_RC);
            }
        }
    }

    public void onActivityResult(int requestCode, int resultCode) {
        mPermHandler.onActivityResult(requestCode);

        if (requestCode == TAKE_PHOTO_RC) {
            File photo = new File(mPhotoPath);
            Context context = mFragment.get().getContext();
            if (resultCode == Activity.RESULT_OK) {
                try {
                    photo = IoUtilsKt.unhide(photo);
                } catch (IOException e) {
                    throw new IllegalStateException(e);
                }

                Uri contentUri = Uri.fromFile(photo);

                mTeam.setHasCustomMedia(true);
                mTeam.setShouldUploadMediaToTba(mShouldUploadMediaToTba);
                mTeam.setMedia(contentUri.getPath());
                mListener.get().onSuccess(mTeam);

                // Tell gallery that we have a new photo
                Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
                mediaScanIntent.setData(contentUri);
                context.sendBroadcast(mediaScanIntent);
            } else {
                photo.delete();
            }
        }
    }

    private File createImageFile(File mediaFolder) {
        return IoUtilsKt.hide(createFile(mTeam.toString(),
                                         "jpg",
                                         mediaFolder,
                                         String.valueOf(System.currentTimeMillis())));
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        mPermHandler.onRequestPermissionsResult(
                requestCode, permissions, grantResults);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(mTeam, flags);
        dest.writeString(mPhotoPath);
        dest.writeInt(mShouldUploadMediaToTba ? 1 : 0);
    }
}

File: app/src/main/java/com/supercilex/robotscouter/util/ui/TeamMediaCreator.kt
package com.supercilex.robotscouter.util.ui

import android.Manifest
import android.app.Activity
import android.arch.lifecycle.LiveData
import android.content.Intent
import android.os.Bundle
import android.provider.MediaStore
import android.support.v4.app.Fragment
import android.support.v4.content.FileProvider
import android.view.Gravity
import androidx.core.net.toUri
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.RobotScouter
import com.supercilex.robotscouter.data.model.Team
import com.supercilex.robotscouter.util.CrashLogger
import com.supercilex.robotscouter.util.asLifecycleReference
import com.supercilex.robotscouter.util.data.ViewModelBase
import com.supercilex.robotscouter.util.data.hidden
import com.supercilex.robotscouter.util.data.ioPerms
import com.supercilex.robotscouter.util.data.mediaFolder
import com.supercilex.robotscouter.util.data.unhide
import com.supercilex.robotscouter.util.logFailures
import com.supercilex.robotscouter.util.logTakeMedia
import com.supercilex.robotscouter.util.providerAuthority
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import org.jetbrains.anko.longToast
import org.jetbrains.anko.runOnUiThread
import pub.devrel.easypermissions.EasyPermissions
import java.io.File
import java.io.IOException
import java.util.Calendar
import java.util.Collections

interface CaptureTeamMediaListener {
    fun startCapture(shouldUploadMediaToTba: Boolean)
}

class TeamMediaCreator : ViewModelBase<Pair<PermissionRequestHandler, Bundle?>>(),
        OnActivityResult, Saveable {
    private val _onMediaCaptured = SingleLiveEvent<Team>()
    val onMediaCaptured: LiveData<Team> get() = _onMediaCaptured
    lateinit var team: Team

    private lateinit var handler: PermissionRequestHandler
    private var photoFile: File? = null
    private var shouldUploadMediaToTba: Boolean? = null

    override fun onCreate(args: Pair<PermissionRequestHandler, Bundle?>) {
        handler = args.first
        args.second?.let {
            photoFile = it.getSerializable(PHOTO_FILE_KEY) as File?
            shouldUploadMediaToTba = it.getBoolean(PHOTO_SHOULD_UPLOAD)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.apply {
            putSerializable(PHOTO_FILE_KEY, photoFile)
            shouldUploadMediaToTba?.let { putBoolean(PHOTO_SHOULD_UPLOAD, it) }
        }
    }

    fun capture(host: Fragment, shouldUploadMediaToTba: Boolean? = null) {
        shouldUploadMediaToTba?.let { this.shouldUploadMediaToTba = it }

        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (takePictureIntent.resolveActivity(RobotScouter.packageManager) == null) {
            RobotScouter.longToast(R.string.fui_error_unknown)
            return
        }
        if (!EasyPermissions.hasPermissions(RobotScouter, *handler.perms.toTypedArray())) {
            handler.requestPerms(host, R.string.media_write_storage_rationale)
            return
        }

        team.logTakeMedia()

        val ref = host.asLifecycleReference()
        async(UI) {
            val file = async {
                try {
                    File(
                            mediaFolder ?: throw IOException("Couldn't create folder"),
                            "${team}_${System.currentTimeMillis()}.jpg"
                    ).hidden().apply {
                        if (!createNewFile()) throw IOException("Unable to create temporary file")
                    }
                } catch (e: IOException) {
                    CrashLogger.onFailure(e)
                    RobotScouter.runOnUiThread { longToast(e.toString()) }
                    null
                }
            }.await() ?: return@async

            photoFile = file
            val photoUri = FileProvider.getUriForFile(RobotScouter, providerAuthority, file)
            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoUri)
            ref().startActivityForResult(takePictureIntent, TAKE_PHOTO_RC)

            if (this@TeamMediaCreator.shouldUploadMediaToTba == true) {
                RobotScouter.longToast(R.string.media_upload_reminder)
                        .setGravity(Gravity.CENTER, 0, 0)
            }
        }.logFailures()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode != TAKE_PHOTO_RC) return

        val photoFile = photoFile!!
        if (resultCode == Activity.RESULT_OK) {
            async(UI) {
                val contentUri = async { photoFile.unhide()?.toUri() }.await()
                if (contentUri == null) {
                    RobotScouter.longToast(R.string.fui_error_unknown)
                    return@async
                }

                _onMediaCaptured.value = team.copy().apply {
                    media = contentUri.path
                    hasCustomMedia = true
                    shouldUploadMediaToTba = this@TeamMediaCreator.shouldUploadMediaToTba!!
                    mediaYear = Calendar.getInstance().get(Calendar.YEAR)
                }

                // Tell gallery that we have a new photo
                RobotScouter.sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE).apply {
                    this.data = contentUri
                })
            }.logFailures()
        } else {
            async { photoFile.delete() }.logFailures()
        }
    }

    companion object {
        private const val TAKE_PHOTO_RC = 334
        private const val PHOTO_FILE_KEY = "photo_file_key"
        private const val PHOTO_SHOULD_UPLOAD = "photo_should_upload"

        val perms: List<String> = Collections.unmodifiableList(ioPerms.toMutableList().apply {
            this += Manifest.permission.CAMERA
        })
    }
}
